nbconvert/exporters/templateexporter.py

```diff
@@ -560,18 +560,37 @@ def get_template_names(self):
                 template_dir = os.path.join(base_dir, base_template)
                 if os.path.exists(template_dir):
                     found_at_least_one = True
-                conf_file = os.path.join(template_dir, 'conf.json')
-                if os.path.exists(conf_file):
-                    with open(conf_file) as f:
-                        conf = recursive_update(json.load(f), conf)
-            for root_dir in root_dirs:
-                template_dir = os.path.join(root_dir, 'nbconvert', 'templates', base_template)
-                if os.path.exists(template_dir):
-                    found_at_least_one = True
-                conf_file = os.path.join(template_dir, 'conf.json')
-                if os.path.exists(conf_file):
-                    with open(conf_file) as f:
-                        conf = recursive_update(json.load(f), conf)
+                    conf_file = os.path.join(template_dir, 'conf.json')
+                    if os.path.exists(conf_file):
+                        with open(conf_file) as f:
+                            conf = recursive_update(json.load(f), conf)
+                    break
+            if not found_at_least_one:
+                for root_dir in root_dirs:
+                    template_dir = os.path.join(root_dir, 'nbconvert', 'templates', base_template)
+                    if os.path.exists(template_dir):
+                        found_at_least_one = True
+                        conf_file = os.path.join(template_dir, 'conf.json')
+                        if os.path.exists(conf_file):
+                            with open(conf_file) as f:
+                                conf = recursive_update(json.load(f), conf)
+                        break
+            if not found_at_least_one:
+                for root_dir in root_dirs:
+                    compatibility_name = base_template + '.tpl'
+                    compatibility_path = os.path.join(root_dir, 'nbconvert', 'templates', 'compatibility', compatibility_name)
+                    if os.path.exists(compatibility_path):
+                        found_at_least_one = True
+                        warnings.warn(
+                            f"5.x template name passed '{self.template_name}'. Use 'lab' or 'classic' for new template usage.",
+                            DeprecationWarning)
+                        self.template_file = compatibility_name
+                        # Without these added the templates can't find their transitive extensions
+                        if base_template == 'basic':
+                            template_names.extend(['classic', 'base'])
+                        if base_template == 'full':
+                            template_names.extend(['lab', 'base'])
+                        break
```

Comment:
This approach may collide with an actual template called `full` or `basic`.

Maybe we should condition on this at a higher level...

Reply:
I was thinking of changing the include paths inside template files to include the template name. I think that would resolve a few issues and not require these hacks for the backwards compatibility, as template extensions are similarly running into 2nd hop extension path failures.

Reply:
Updated PR with the change I discussed -- extending existing templates from custom templates no longer explodes when files are included across template boundaries. e.g. Fixes https://github.com/jupyter/nbconvert/issues/1394

Reply:
> changing the include paths inside template files to include the template name

I am not sure about this, because it is a backward incompatible change with respect to the current behavior.

Reply:
How is it backwards incompatible? The templates behave identically within nbconvert but now when extended from external templates will behave correctly (since they don't have the proper template paths set to include this template). It also protects from multiple inheritance of file extensions -- e.g. there's ambiguity when importing null.js is some places if you didn't include the template name in the extension path.

Reply:
It is backward incompatible for custom template authors who inherit from base.